sstables: add options for disabling long-term index caching

michoecho opened the pull request:
A draft of a potential fix for #11202. See that issue and commit messages `sstables: add a flag for disabling long-term index caching` and `sstables: index_entry: add use_caching::cache_locally` for some discussion.

`sstables: add a flag for disabling long-term index caching` alone is enough as a band-aid fix. If I understood the history of this code correctly, enabling the flag added in that patch should revert index caching to it's state at 4.5.

But since in some cases (partitions with promoted index, especially if it fits on a single page with the partition index entry) disabling index file caching is likely to lead to additional disk reads or (w.r.t. 4.6), I also added a mode in between full caching and BYPASS CACHE, which caches index file pages only for the lifetime of the index reader. It didn't come out that small, though, so I'm not sure if it's a good fit for backporting.

Passing the flag around in this series is hacked in with a global variable. But there is no obstacle to passing it around normally or making it per-table, if needed.

avikivity commented on db/config.cc:
```diff
@@ -901,6 +901,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
         "Ignore truncation record stored in system tables as if tables were never truncated.")
     , force_schema_commit_log(this, "force_schema_commit_log", value_status::Used, false,
         "Use separate schema commit log unconditionally rater than after restart following discovery of cluster-wide support for it.")
+    , cache_index_pages(this, "cache_index_pages", value_status::Used, true,
```
Let's make this LiveUpdate so a rolling restart can be avoided.

avikivity commented on main.cc:
```diff
@@ -560,6 +560,10 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
+            smp::invoke_on_all([cache_index_pages = cfg->cache_index_pages()] {
+                sstables::global_cache_index_pages = cache_index_pages;
+            }).get();
```
This won't work with LiveUpdate, unless we add a listener and re-broadcast it, or have the config accessed each time.

(or fix #7316, really overdue)

avikivity commented:
Looks good. @tgrabiec please ack the strategy in the last patch.

tgrabiec commented on sstables/index_reader.hh:
```diff
@@ -441,6 +444,13 @@ class index_reader {
+        return _local_cached_file;
+    }
+
     seastar::shared_ptr<cached_file> get_cached_file() {
```
I think get_cached_file() should use _local_cached_file if it's available, so that we don't have two separate caches for the partition index and bsearch cursor.

michoecho replied:
Yes, good catch. It was supposed to, I must have lost the hunk during cleanup/rebase.

tgrabiec commented:
We could compare perf_fast_forward results on 4.5 to catch potential regressions due to things we have missed.

michoecho commented:
Also, note on testing: I tested the changes manually by tracing the queries (`TRACING ON`, not the PT kind, though that would have worked too) and observing the I/O. I will try `perf_fast_forward` too.

michoecho commented:
v2:
- Modified `sstables: index_entry: add use_caching::cache_locally` to make the bsearch cursor use the same cached file as the partition index.
- Modified `sstables: add a flag for disabling long-term index caching` to make the flag live updatable.
- Modified both of the above by adding comments to the global variable.

michoecho commented:
@avikivity I made the flag live-updateable. Please re-review the first commit (`sstables: add a flag for disabling long-term index caching`).

tgrabiec commented on main.cc:
```diff
@@ -569,6 +569,12 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            // We pass this piece of config through a global as a temporary hack.
+            // See the comment at the definition of sstables::global_cache_index_pages.
+            smp::invoke_on_all([&cfg] {
+                sstables::global_cache_index_pages = cfg->cache_index_pages.operator utils::updateable_value<bool>();
```
Is this operator call necessary? It's not marked explicit.

michoecho replied:
It's not necessary per se, but I wanted to make explicit the fact that we are dealing with `updateable_value<bool>` rather than a `bool`.
E.g. if you typed `sstables::global_cache_index_pages = cfg->cache_index_pages;` it would work as intended, because `cfg->cache_index_pages` (a `named_value<bool>`) is implicitly convertible to `updateable_value<bool>`, but you could very easily confuse this with `sstables::global_cache_index_pages = cfg->cache_index_pages();`. `cfg->cache_index_pages()` is a `bool` and `updateable_value<T>` has `operator=(T)`, so this would still compile and set the flag correctly on startup, but would break the live updating silently.

tgrabiec replied:
A standard way to do this would be via static_cast<>() rather than calling the operator.

The problem you point out seems to be at the API level, so perhaps needs an API-level solution if we acknowledge it. Those explicit operator calls are ugly. We could add `named_value::updateable_value()` member method for the conversion.

michoecho commented:
Abandoning the draft. https://github.com/scylladb/scylladb/issues/11202 was relieved with a simpler version.
